src/query_plan/fetch_dependency_graph.rs

```diff
@@ -292,18 +634,568 @@ impl DeferTracking {
+        let (required_node_id, require_path) = handle_requires(
+            dependency_graph,
+            edge_id,
+            conditions,
+            (stack_item.node_id, &stack_item.node_path),
+            stack_item.context,
+            &updated.defer_context,
+            created_nodes,
+        )?;
+        updated.node_id = required_node_id;
+        updated.node_path = require_path;
+    }
+    if let OpPathElement::Field(field) = &updated_operation {
+        if *field.data().name() == TYPENAME_FIELD {
+            // Because of the optimization done in `QueryPlanner.optimizeSiblingTypenames`,
+            // we will rarely get an explicit `__typename` edge here.
+            // But one case where it can happen is where an @interfaceObject was involved,
+            // and we had to force jumping to another subgraph for getting the "true" `__typename`.
+            // However, this case can sometimes lead to fetch dependency node
+            // that only exists for that `__typename` resolution and that "looks" useless.
+            // That is, we could have a fetch dependency node that looks like:
+            // ```
+            //   Fetch(service: "Subgraph2") {
+            //     {
+            //       ... on I {
+            //         __typename
+            //         id
+            //       }
+            //     } =>
+            //     {
+            //       ... on I {
+            //         __typename
+            //       }
+            //     }
+            //   }
+            // ```
+            // but the trick is that the `__typename` in the input
+            // will be the name of the interface itself (`I` in this case)
+            // but the one return after the fetch will the name of the actual implementation
+            // (some implementation of `I`).
+            // *But* we later have optimizations that would remove such a node,
+            // on the node that the output is included in the input,
+            // which is in general the right thing to do
+            // (and genuinely ensure that some useless nodes created when handling
+            // complex @require gets eliminated).
+            // So we "protect" the node in this case to ensure
+            // that later optimization doesn't kick in in this case.
+            let updated_node = FetchDependencyGraph::node_weight_mut(
+                &mut dependency_graph.graph,
+                updated.node_id,
+            )?;
+            updated_node.must_preserve_selection_set = true
+        }
+    }
+    let edge = child.tree.graph.edge_weight(edge_id)?;
+    if let QueryGraphEdgeTransition::InterfaceObjectFakeDownCast { .. } = &edge.transition {
+        // We shouldn't add the operation "as is" as it's a down-cast but we're "faking it".
+        // However, if the operation has directives, we should preserve that.
+        let OpPathElement::InlineFragment(inline) = updated_operation else {
```

Comment:
I think JS `FragmentElement` represents both inline fragment and named spreads -> their selections actually reference same `FragmentElement` (https://github.com/apollographql/federation/blob/main/internals-js/src/operations.ts#L3122)

Reply:
Here `update_operation` has type:

```rust
pub(crate) enum OpPathElement {
    Field(NormalizedField),
    InlineFragment(NormalizedInlineFragment),
}
```

There is no variant for named fragments, so presumably they’ve been inlined when or before creating an `OpPathElement`.